### ISSUE TITLE
What exception should throw in v2 and v3

### ISSUE THREAD

**Issue by svycka** [zendframework/zend-servicemanager#88]
most if not all modules throws incorrect exception not this https://github.com/zendframework/zend-servicemanager/blob/release-2.7/src/AbstractPluginManager.php#L126 examples: 
* https://github.com/zendframework/zend-inputfilter/blob/8c3bb64f72d4014edf7b53641d017a6b5340a577/src/InputFilterPluginManager.php#L85
* https://github.com/zendframework/zend-hydrator/blob/76a85791a82f8f832d3346aa47bcce8797268851/src/HydratorPluginManager.php#L68
* and others..

is this correct behavior?
if yes then now in v3 throwing `Zend\ServiceManager\InvalidServiceException` can be BC break?

**Comment by kynx** [comment]
@svycka I've added some stuff about exceptions to the migration docs in #87 - I'd appreciate any feedback.

TL;DR
* v2 should continue to throw the old exception
* v3 should throw an `InvalidServiceException` (or an exception extending it)

If it's possible to change the old exception to extend `InvalidServiceException` while preserving BC I guess that would be the way to go. But as you note in lots of cases it's package-specific but generic (ie `Exception\RuntimeException`) so I guess BC breakage is inevitable.

**Comment by weierophinney** [comment]
Closing; #93 and #96 allow writing tests for this behavior now, and #93 provides a methodology for throwing different exceptions in v2 usage vs v3 usage.

We will be updating the various plugin managers over the coming weeks.
